modules/packages/LinearAlgebra.chpl

```diff
@@ -1270,54 +1269,135 @@ proc det (A: [?Adom] ?eltType) {
   var (LU,ipiv,numSwap) = _lu(A);
   const pdet = if numSwap % 2 == 0 then 1 else -1;
 
-  // L[i,i] always = 1, so we only need to take the 
+  // L[i,i] always = 1, so we only need to take the
   // diagonal product of U
 
   return (* reduce [i in Adom.dim(0)] LU[i,i]) * pdet;
 }
 
-/* Return the solution ``x`` to the linear system `` L * x = b `` 
+/*
+  Compute the default norm on `x`.
+
+  For a 1D array this is the 2-norm, for a 2D array, this is the Frobenius
+  norm.
+
+  :rtype: x.eltType
+*/
+proc norm(x: [], param p = normType.default) {
+  if x.rank > 2 {
+    compilerError("Norms not implemented for array ranks > 2D");
+  }
+
+  if p == normType.default {
+    param defaultType: normType = if x.rank == 1 then normType.norm2 else normType.normFrob;
+    return norm(x, defaultType);
+  } else {
+    return _norm(x, p);
+  }
+}
+
+/*
+  Indicates the different types of norms supported by :proc:`norm`:
+
+    * Default - depends on array dimensions. See :proc:`norm` for details.
+    * 1-norm
+    * 2-norm
+    * Infinity norm
+    * Frobenius norm
+ */
+enum normType {
+  default,
+  norm1,
+  norm2,
+  normInf,
+  normFrob
+};
+
+/*
+  Compute the norm indicated by `p` on the 1D array `x`.
+
+  :rtype: x.eltType
+ */
+private proc _norm(x: [], param p: normType) where x.rank == 1 {
+  select (p) {
+  when normType.norm1 do return + reduce abs(x);
+  when normType.norm2 do return sqrt(+ reduce (abs(x)*abs(x)));
+  when normType.normInf do return max reduce abs(x);
+  when normType.normFrob do return sqrt(+ reduce (abs(x)*abs(x)));
+  otherwise halt("Unexpected norm type");
+  }
+}
+
+/*
+  Compute the norm indicated by `p` on the 2D array `x`.
+
+  `p` cannot be `normType.norm2`.
+
+  :rtype: x.eltType
+ */
+proc _norm(x: [?D], param p: normType) where x.rank == 2 {
+  select (p) {
+  when normType.norm1 do
+    return max reduce forall j in D.dim(1) do (+ reduce abs(x[D.dim(0), j..j]));
+
+  when normType.norm2 {
+    halt("2-norm for 2D arrays are not yet implemented");
```

Comment:
Can't we make this a compile-time error?

Reply:
Ah yes, good point!